Framework/API/src/SpectrumInfo.cpp

```diff
@@ -82,36 +79,50 @@ double SpectrumInfo::signedTwoTheta(const size_t index) const {
     throw std::logic_error(
         "Two theta (scattering angle) is not defined for monitors.");
 
-  // Note: This function has big overlap with the method
-  // MatrixWorkspace::detectorSignedTwoTheta(). The plan is to eventually remove
-  // the latter, once SpectrumInfo is in widespread use.
-  const Kernel::V3D samplePos = samplePosition();
-  const Kernel::V3D beamLine = samplePos - sourcePosition();
-
-  if (beamLine.nullVector()) {
-    throw Kernel::Exception::InstrumentDefinitionError(
-        "Source and sample are at same position!");
+  double signedTwoTheta{0.0};
+  const auto &dets = getDetectorVector(index);
+  for (const auto &det : dets) {
+    const auto &detIndex = m_detectorInfo.indexOf(det->getID());
+    m_detectorInfo.setCachedDetector(detIndex, det);
+    signedTwoTheta += m_detectorInfo.signedTwoTheta(detIndex);
   }
-  // Get the instrument up axis.
-  const Kernel::V3D &instrumentUpAxis =
-      m_instrument->getReferenceFrame()->vecPointingUp();
-  return getDetector(index)
-      .getSignedTwoTheta(samplePos, beamLine, instrumentUpAxis);
+  return signedTwoTheta / static_cast<double>(dets.size());
 }
 
 /// Returns the position of the spectrum with given index.
 Kernel::V3D SpectrumInfo::position(const size_t index) const {
-  return getDetector(index).getPos();
+  // The body of this function is almost a 1:1 duplicate of
+  // DetectorGroup::getPos. The former should be cleaned up once subsequent
+  // changes allow for this, and the latter should be removed.
+  Kernel::V3D newPos;
+  const auto &dets = getDetectorVector(index);
+  for (const auto &det : dets) {
+    const auto &detIndex = m_detectorInfo.indexOf(det->getID());
+    m_detectorInfo.setCachedDetector(detIndex, det);
+    newPos += m_detectorInfo.position(detIndex);
```

Comment:
So I think this bit of code answers one of my questions - `DetectorInfo` can be accessed in a mutable fashion. How does the `SpectrumInfo` stay synchronized. We always compute from `DetectorInfo`.